### PR TITLE
Re-add Vanished players don't have rights patch

### PR DESCRIPTION
**Pull request by hugmanrique**
Readds patch 0177. Moves away from `AxisAlignedBB`s to `VoxelShape`s.

The `checkNoVisiblePlayerCollisions` method is similar to the `a(Entity, VoxelShape)` method.

The `defaultReturn` boolean was adapted to use the `VoxelShape` provided by `IBlockData#h(World, BlockPosition)`, which is also used in the `EntityArrow` class (which I used as inspiration). Here's the changes on that class:

1.12:
```java
if (iblockdata.getMaterial() != Material.AIR) {
        AxisAlignedBB axisalignedbb = iblockdata.d(this.world, blockposition); // Moved to VoxelShape

        if (axisalignedbb != Block.k && axisalignedbb.a(blockposition).b(new Vec3D(this.locX, this.locY, this.locZ))) {
            this.inGround = true;
        }
}
```

1.13:
```java
if (!iblockdata.isAir() && !flag) {
        VoxelShape voxelshape = iblockdata.h(this.world, blockposition);

        if (!voxelshape.b()) {
            // Other code
           }
        }
}
```

### CODE REVIEW

File: Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+     }
+ 
++    // Paper start - Based on method below
++    /**
++     * @param entity causing the action ex. block placer
++     * @param voxelshape area to search within
++     * @return if there are no visible players colliding
++     */
++    public boolean checkNoVisiblePlayerCollisions(@Nullable Entity entity, VoxelShape voxelshape) {
++        if (voxelshape.b()) {
```
Comment (kashike) [inline Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch]: Please create an obfuscation helper for `b()`.

File: Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
++            return true;
++        } else {
++            List list = this.getEntities((Entity) null, voxelshape.a());
```
Comment (kashike) [inline Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch]: Please create an obfuscation helper for `a()`.

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+index 1cecccef..09e922be 100644
+--- a/src/main/java/net/minecraft/server/ItemBlock.java
++++ b/src/main/java/net/minecraft/server/ItemBlock.java
+@@ -70,7 +70,8 @@ public class ItemBlock extends Item {
+ 
+     protected boolean b(BlockActionContext blockactioncontext, IBlockData iblockdata) {
+         // CraftBukkit start - store default return
+-        boolean defaultReturn = iblockdata.canPlace(blockactioncontext.getWorld(), blockactioncontext.getClickPosition()) && blockactioncontext.getWorld().a(iblockdata, blockactioncontext.getClickPosition());
++        final World world = blockactioncontext.getWorld(); // Paper
++        boolean defaultReturn = iblockdata.canPlace(world, blockactioncontext.getClickPosition()) && world.a(iblockdata, blockactioncontext.getClickPosition()) && world.checkNoVisiblePlayerCollisions(blockactioncontext.getEntity(), iblockdata.h(world, blockactioncontext.getClickPosition())); // Paper - Use our entity search
```
Comment (aikar) [inline Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch]: Please also add an obf helper for iblockdata.h() here too, as that method was not originally in the line replaced.

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+     private Entity ax;
+diff --git a/src/main/java/net/minecraft/server/IBlockData.java b/src/main/java/net/minecraft/server/IBlockData.java
+index c8f305e6..d3103cb4 100644
+--- a/src/main/java/net/minecraft/server/IBlockData.java
++++ b/src/main/java/net/minecraft/server/IBlockData.java
+@@ -183,6 +183,8 @@ public interface IBlockData extends IBlockDataHolder<IBlockData> {
+         return this.getBlock().f(this, iblockaccess, blockposition);
+     }
+ 
++    default VoxelShape getBlockShape(IBlockAccess iblockaccess, BlockPosition blockposition) { return h(iblockaccess, blockposition); } // Paper - OBFHELPER
```
Comment (kashike) [inline Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch]: Please put this on the line *directly* (no blank line between) above the method the obfuscation helper is for.
Example:
```java
default VoxelShape getBlockShape(IBlockAccess iblockaccess, BlockPosition blockposition) { return h(iblockaccess, blockposition); } // Paper - OBFHELPER
// <h(IBlockAccess, BlockPosition) method on this line>
```

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+         blockactioncontext.getWorld().getServer().getPluginManager().callEvent(event);
+diff --git a/src/main/java/net/minecraft/server/VoxelShape.java b/src/main/java/net/minecraft/server/VoxelShape.java
+index 53c9f218..e1f96178 100644
+--- a/src/main/java/net/minecraft/server/VoxelShape.java
++++ b/src/main/java/net/minecraft/server/VoxelShape.java
+@@ -30,6 +30,8 @@ public abstract class VoxelShape {
+         }
+     }
+ 
++    public AxisAlignedBB getBounds() { return a(); } // Paper - OBFHELPER
```
Comment (aikar) [inline Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch]: This Obf helper is also in the wrong location, see kashikes comment.

File: Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
++
+     protected double a(EnumDirection.EnumAxis enumdirection_enumaxis, int i) {
+         return i < 0 ? -1.7976931348623157E308D : (i > this.a.c(enumdirection_enumaxis) ? Double.MAX_VALUE : this.b(enumdirection_enumaxis, i));
+     }
+@@ -44,6 +46,8 @@ public abstract class VoxelShape {
+         return this.a.a();
+     }
+ 
++    public boolean isEmpty() { return b(); } // Paper - OBFHELPER
```
Comment (aikar) [inline Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch]: This Obf helper is also in the wrong location, see kashikes comment.

File: Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
++++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+@@ -804,6 +804,13 @@ public class CraftEventFactory {
+         Projectile projectile = (Projectile) entity.getBukkitEntity();
+         org.bukkit.entity.Entity collided = position.entity.getBukkitEntity();
+         com.destroystokyo.paper.event.entity.ProjectileCollideEvent event = new com.destroystokyo.paper.event.entity.ProjectileCollideEvent(projectile, collided);
++
++        if (projectile.getShooter() instanceof Player && collided instanceof Player) {
++            if (!((Player) projectile.getShooter()).canSee((Player) collided)) {
++                event.setCancelled(true);
```
Comment (aikar) [inline Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch:119]: Go ahead and return here. We shouldn't fire the event. There should not be opportunity for a plugin to break this.

Comment (aikar) [reply Spigot-Server-Patches/0307-Vanished-players-don-t-have-rights.patch:119]: I know original patch didn't do that, time to improve ;)

**Review by aikar — CHANGES_REQUESTED** [review]

**Review by aikar — APPROVED** [review]

**Comment by aikar** [comment]
are you able to resolve the conflicts or do you need us to?

**Pull request merged**